File: abc/abc129/abc129_c.cs
// https://atcoder.jp/contests/abc129/submissions/8303995

using System;
using System.Linq;

namespace CSharp
{
    class Program
    {
        static readonly int mod = 1000000007;

        public static void Main()
        {
            var input = Console.ReadLine().Split(' ').Select(int.Parse).ToArray();
            var N = input[0];
            var M = input[1];

            var a = new int[M];
            for (var i = 0; i < M; i++)
            {
                a[i] = int.Parse(Console.ReadLine());
            }

            var dp = new int[N + 1];
            dp[0] = 1;
            var index = 0;
            for (var i = 1; i <= N; i++)
            {
                if (index < M && a[index] == i)
                {
                    dp[i] = 0;
                    index++;
                }
                else
                {
                    dp[i] = (dp[i - 1] + (i != 1 ? dp[i - 2] : 0)) % mod;
                }
            }
            Console.WriteLine(dp[N]);
        }
    }
}

File: abc/abc129/abc129_c.fs
// https://atcoder.jp/contests/abc129/submissions/12152311

let MOD = 1000000007

let [| N; M |] = stdin.ReadLine().Split() |> Array.map int

let a =
    [| for _ in [ 1..M ] do
        yield (stdin.ReadLine() |> int) |]

let dp = Array.zeroCreate<int> (N + 1)

dp.[0] <- 1

let rec resolver idx i =
    if i = N + 1 then ()
    elif idx < M && a.[idx] = i then
        dp.[i] <- 0
        resolver (idx + 1) (i + 1)
    else
        dp.[i] <- (dp.[i - 1] + if i <> 1 then dp.[i - 2]
                                else 0)
                  % MOD
        resolver idx (i + 1)

resolver 0 1

dp.[N] |> stdout.WriteLine
